storage/innobase/trx/trx0trx.cc

```diff
@@ -2468,3 +2507,63 @@ trx_set_rw_mode(
 		trx->read_view.set_creator_trx_id(trx->id);
 	}
 }
+
+/*************************************************************//**
+Return difference in uncommitted count for a single undo record. */
+ib_int64_t get_diff_from_rec(
+/*============*/
+    trx_undo_rec_t* undo_rec,		/*!< in: undo record */
+    table_id_t table_id)           	/*!< in: table ID */
+{
+	ulint type, cmpl_info;
+	bool updated_extern;
+	undo_no_t undo;
+	table_id_t rec_table_id;
+
+	trx_undo_rec_get_pars(undo_rec, &type, &cmpl_info, &updated_extern, &undo,
+		&rec_table_id);
+
+	if (rec_table_id != table_id)
+		return 0;
+
+	switch (type) {
+		case TRX_UNDO_INSERT_REC:
+			return 1;
+		case TRX_UNDO_UPD_DEL_REC:
+		case TRX_UNDO_DEL_MARK_REC:
+			return -1;
+		default:
+			return 0;
+	}
+}
+
+/*************************************************************//**
+Return number of uncommitted records for table within transaction
+@param[in]	table 	table to count uncommitted records for
+*/
+ib_int64_t trx_t::uncommitted_count(dict_table_t* table)
+{
+	trx_undo_t* undo;
+	trx_undo_rec_t* undo_rec;
+	mtr_t mtr;
+	ib_int64_t count = 0;
+
+	undo = rsegs.m_redo.undo;
+
+	if (undo) {
+		mtr_start(&mtr);
+
+		undo_rec = trx_undo_get_first_rec(
+			undo->rseg->space, undo->hdr_page_no, undo->hdr_offset,
+			RW_S_LATCH, &mtr);
+
+		while (undo_rec) {
+			count += get_diff_from_rec(undo_rec, table->id);
+			undo_rec = trx_undo_get_next_rec(undo_rec, undo->hdr_page_no,
+				undo->hdr_offset, &mtr);
+		}
+		mtr_commit(&mtr);
+	}
```

Comment:
Here we seem to be looping over all undo log records over all tables that the transaction has touched. Could we rather update the counts for all touched tables by looping over the undo log records only once?

Also, undo log records for a transaction can be spanning multiple pages. Here, we only seem to be iterating the records in the first undo page only.

Reply:
>Here we seem to be looping over all undo log records over all tables that the transaction has touched. Could we rather update the counts for all touched tables by looping over the undo log records only once?

Updated `trx_update_persistent_counts()` to only loop over the undo records once.

>Also, undo log records for a transaction can be spanning multiple pages. Here, we only seem to be iterating the records in the first undo page only.

Seems like `trx_undo_get_next_rec()` iterates through multiple pages?
```
trx_undo_rec_t*
trx_undo_get_next_rec(
/*==================*/
	trx_undo_rec_t*	rec,	/*!< in: undo record */
	ulint		page_no,/*!< in: undo log header page number */
	ulint		offset,	/*!< in: undo log header offset on page */
	mtr_t*		mtr)	/*!< in: mtr */
{
	ulint		space;
	trx_undo_rec_t*	next_rec;

	next_rec = trx_undo_page_get_next_rec(rec, page_no, offset);

	if (next_rec) {
		return(next_rec);
	}

	space = page_get_space_id(page_align(rec));

	return(trx_undo_get_next_rec_from_next_page(space,
						    page_align(rec),
						    page_no, offset,
						    RW_S_LATCH, mtr));
}
```